BoundaryCondition class for GH

prayush opened the pull request:
## Proposed changes

This PR adds a `BoundaryCondition` derived class to apply Bjorhus-type corrections to time derivatives of GH variables on outer domain boundaries.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

~~This PR depends on 3 others (#2998 #2999 #3000 ) and includes those as the first 3 commits, and the 4th commit also only adds corrections to those 3. None of the first 4 commits should be reviewed here! Please review from 5th commit onward (int his discussion, my commit ordering is simply chronological).~~

~~The first commit in this PR comes from #3000 . Please ignore it while reviewing.~~

prayush committed:
Add Python functions to compute lapse, shift, extrinsic curv

moxcodes commented on tests/Unit/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.py:
```diff
@@ -131,6 +131,12 @@ def spacetime_deriv_detg(sqrt_det_spatial_metric, inverse_spatial_metric,
 
 
+def extrinsic_curvature(spacetime_normal_vector, pi, phi):
+    return 0.5 * pi[1:, 1:] + 0.5 * np.einsum(
```
Suggest parens to make the formatting less odd.

prayush replied:
added!

moxcodes commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/Bjorhus.hpp:
```diff
@@ -0,0 +1,230 @@
+/*!
+ * \brief Sets constraint preserving boundary conditions using the Bjorhus
+ * method.
+ */
```
It would be helpful to provide in this documentation a link to the Bjorhus paper that describes the general procedure, any available link to GR papers about the procedure applied to GH, and code references to the individual implementation functions that have all of the mathematical details for setting the respective components of the characteristic fields.

prayush replied:
Added a more descriptive summary with useful references.

moxcodes commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/Bjorhus.cpp:
```diff
@@ -0,0 +1,488 @@
+void set_bc_corr_zero_when_char_speed_is_positive(
+    const gsl::not_null<T*> dt_v_corr,
+    const DataVector& char_speed_u) noexcept {
+  auto it = dt_v_corr->begin();
```
can this declaration be the initialization entry of the `for`?

prayush replied:
yup, moved it there

moxcodes commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/Bjorhus.cpp:
```diff
@@ -0,0 +1,488 @@
+      make_not_null(&bc_dt_v_psi), char_speeds.at(0));
+  helpers::set_bc_corr_zero_when_char_speed_is_positive(
+      make_not_null(&bc_dt_v_zero), char_speeds.at(1));
+  helpers::set_bc_corr_zero_when_char_speed_is_positive(
```
Is there some mathematical guarantee that means that v_plus doesn't need this check also?

prayush replied:
The characteristic speed of `v_plus` is `lapse - normal _k shift^k`. The hope here is that the external boundary is sufficiently far away, such that `lapse --> 1` and  `shift^k --> 0`, and also that the mesh velocity is not close to being luminal. This would fail if, of course, the boundary is very close to the source, but then so will this whole treatment of the boundary conditions anyway.

moxcodes commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/Bjorhus.cpp:
```diff
@@ -0,0 +1,488 @@
+      make_not_null(&bc_dt_v_minus), char_speeds.at(3));
+
+  // Convert corrections to dt<evolved variables>
+  auto dt_evolved_vars = evolved_fields_from_characteristic_fields(
```
Do I understand this correctly that this call is just passing the time derivatives in place of the characteristics and using the result as the time derivatives of the evolved fields?
Is there any danger of other spacetime-dependent factors that have nonvanishing time derivatives that give rise to additional corrections to this calculation?

prayush replied:
The corrections are applied to the incoming projections of the time derivatives of evolved vars, and *not* to the time derivatives of incoming characteristic variables. Say the evolved fields are `u^A`, and `D^A_B` is the characteristic projection operator. The Bjorhus corrections were derived as:

`D^A_B \partial_t u^B ---> D^A_B \partial_t u^B + corrections^A`

which could also be written as

`\partial_t u^B ----> \partial_t u^B + (D^{-1})^B_A corrections^A`

The `dt_evolved_vars` here is simply the last term `(D^{-1})^B_A corrections^A`. See the discussion around Eq. (62) of [Lindblom et al 2005](https://arxiv.org/abs/gr-qc/0512093).

geoffrey4444 replied:
Please add a comment of your answer to the code @prayush

markscheel replied:
This explanation of `dt_evolved_vars` should be made into a comment in the code.

moxcodes commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/Bjorhus.cpp:
```diff
@@ -0,0 +1,488 @@
+                        ::Tags::Tempia<0, Dim, Frame::Inertial, DataVector>>>
+      local_buffer(get_size(get<0>(normal_covector)));
+
+  //   auto& lapse = get<::Tags::TempScalar<0, DataVector>>(local_buffer);
```
remove unused line

prayush replied:
removed

moxcodes commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/Bjorhus.cpp:
```diff
@@ -0,0 +1,488 @@
+  auto& two_index_constraint =
+      get<::Tags::Tempia<0, Dim, Frame::Inertial, DataVector>>(local_buffer);
+
+  *inverse_spacetime_metric = determinant_and_inverse(spacetime_metric).second;
```
I think the advice is usually to use the inverse spatial metric to compute the inverse spacetime metric rather than getting it from direct inversion

prayush replied:
makes sense, changed it to avoid inversion

moxcodes commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/Bjorhus.cpp:
```diff
@@ -0,0 +1,488 @@
+                       *inverse_spatial_metric);
+
+  gr::shift(make_not_null(&shift), spacetime_metric, *inverse_spatial_metric);
+  //   gr::lapse(make_not_null(&lapse), shift, spacetime_metric);
```
remove unused

prayush replied:
removed

moxcodes commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/Bjorhus.cpp:
```diff
@@ -0,0 +1,488 @@
+                                     *spacetime_unit_normal_vector,
+                                     *unit_interface_normal_vector);
+
+  const auto dt_char_fields = characteristic_fields(
```
same question about extra time derivatives for tensor factors in the characteristic computaiton

prayush replied:
See [above](https://github.com/sxs-collaboration/spectre/pull/3172#discussion_r646623671)

moxcodes commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/Bjorhus.py:
```diff
@@ -219,3 +222,271 @@ def set_dt_v_minus_constraint_preserving_physical(
+    unit_interface_normal_vector = np.einsum('i,ij->j', normal_covector,
+                                             inverse_spatial_metric)
+    shift = gr.shift(spacetime_metric, inverse_spatial_metric)
+    # lapse = gr.lapse(shift, spacetime_metric)
```
remove unused

prayush replied:
removed

prayush committed:
BoundaryCondition class for Bjorhus-type corrections to GH dt_vars

geoffrey4444 commented:
@prayush please rebase and fix conflicts; then I'll take a look (might take me a bit to get through the whole thing, but I'll do the best I can)

prayush committed:
Add gauge function and its spacetime deriv to tensors returned by GH::DuDt

prayush commented:
Hi @geoffrey4444 - this has been rebased onto develop and can be tested with evolutions.

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/Bjorhus.cpp:
```diff
@@ -0,0 +1,488 @@
+double min_characteristic_speed(
+    const std::array<DataVector, 4>& char_speeds) noexcept {
+  std::array<double, 4> min_speeds{
+      {min(char_speeds.at(0)), min(char_speeds.at(1)), min(char_speeds.at(2)),
```
`at` -> `[]`. `at` does a runtime bounds check, but since you have hard-coded numbers the compiler also does a bounds check with `[]`

prayush replied:
fixed

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/Bjorhus.cpp:
```diff
@@ -0,0 +1,488 @@
+    const gsl::not_null<T*> dt_v_corr,
+    const DataVector& char_speed_u) noexcept {
+  auto it = dt_v_corr->begin();
+  for (; it != dt_v_corr->end(); ++it) {
```
I _think_ this could be a range-based for: `for (DataVector& component : *dt_v_corr)`, which I personally find easier to read and deal with than iterators, but it's not always possible to use :)

prayush replied:
Yup, just my inner-C spoke out. replaced with range-based for

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/Bjorhus.cpp:
```diff
@@ -0,0 +1,488 @@
+        constraint_char_zero_plus, constraint_char_zero_minus, phi, d_phi, d_pi,
+        char_speeds);
+  } else {
+    ERROR("Failed to set dtVMinus. Input option \""
```
did you mean to stream something between the quotes here? Currently this will say `Input option "" Must be one of`.

nilsdeppe replied:
ping

prayush replied:
Not really - this check should be handled much earlier by `convert_constraint_preserving_bjorhus_type_from_yaml` when using options. In fact, this `else {}` clause should never be reached. I have left it in to complete the `if - else if  - ` chain, but can remove it also.

prayush replied:
Fixed the error message by removing the `\"\"` here.

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/Bjorhus.cpp:
```diff
@@ -0,0 +1,488 @@
+      get<gr::Tags::SpacetimeMetric<Dim, Frame::Inertial, DataVector>>(
+          dt_evolved_vars);
+
+  // Subtract out original dt<vars>
```
can this code block be removed or should it be uncommented?

prayush replied:
removed already

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/Bjorhus.cpp:
```diff
@@ -0,0 +1,488 @@
+      local_buffer(get_size(get<0>(normal_covector)));
+
+  //   auto& lapse = get<::Tags::TempScalar<0, DataVector>>(local_buffer);
+  auto& shift =
```
why are you computing these instead of grabbing them from the temporary tags? lapse, shift, spatial metric, spacetime normals, inverse spatial metric, inverse spacetime metric, three index constraint, are all computed in the bulk and could be sliced. Naively I'd expect that, at least things like the inverse spacetime metric and three-index constraint would be better to slice than to recompute. The right thing to do is whatever is fastest, but it's not completely obvious to me in general which is faster, though I assume slicing will win for all except maybe the most trivial tags. If you'd like to defer this to a followup PR, that's fine, but a code comment and an issue would be good

nilsdeppe replied:
ping

prayush replied:
I wasn't sure of which would be faster either. There would be the computation cost and the memory allocation cost. I thought at least in the case when these vars are being recomputed here, the memory allocation can be done in one buffer, while the same won't be possible when slicing from volume. But not having profiled this, I am quite happy to change the behavior here. Pushed a patch that makes use of as many temp tags as available.

nilsdeppe replied:
There's no memory allocation cost for the slicing, it's already paid for just by having an interface. We have one Variables that stores everything on the interface, so switching actually means fewer allocations :)

nilsdeppe reviewed (changes requested):
I've only reviewed the commit `BoundaryCondition class for GH` because it wasn't clear to me which of the other commits are in different PRs

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/TimeDerivative.cpp:
```diff
@@ -71,6 +74,9 @@ void TimeDerivative<Dim>::apply(
   // Need constraint damping on interfaces in DG schemes
   *temp_gamma1 = gamma1;
   *temp_gamma2 = gamma2;
+  // Need guage function on external faces for gauge-constraint preservation
```
typo: `Need guage` -> `Need gauge`

prayush replied:
fixed

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/Bjorhus.cpp:
```diff
@@ -227,25 +226,33 @@ std::optional<std::string> ConstraintPreservingBjorhus<Dim>::dg_time_derivative(
     }
   }
+  // If the chosen boundary lies inside the apparent horizon, return here
+  if (helpers::min_characteristic_speed(char_speeds) >= 0.) {
```
when do we expect to be applying the BC at the horizon? I'm a bit surprised that you need to handle this case since we can now specify different boundary conditions (including `Outflow`) on different boundaries

prayush replied:
You're right in part and its really the help text that has to be corrected (which I have done now). The exception case is when we have a contracting outer boundary and the net incoming char speed for, say, `VMinus` becomes 0. This may not be a realistic case, but that is what this block is supposed to handle. And lets see what it is doing anyway: its checking if all speeds are positive, skip further computation. Even if we remove this box entirely, the same behavior will ensue later on, with the only difference being we'll calculate a lot of things and never use them.

nilsdeppe replied:
okay, that sounds good! I think the comment is clear now :) Thanks!

nilsdeppe reviewed (commented):
LGTM, please go ahead and squash

geoffrey4444 commented:
@prayush Tests are failing

prayush commented:
@nilsdeppe @geoffrey4444 - squashed previous fixups, fixed unit test failures, and squashed that fix as well.

nilsdeppe commented:
@prayush great! Looks like CI found some more failure cases. Maybe it's worth running locally for say 1000 times (ctest can do this automatically) in case you haven't done that to try and catch some more of these types of issues. I'm a bit worried that we might merge this and then every other PR fails in this test :)

prayush commented:
> @prayush great! Looks like CI found some more failure cases. Maybe it's worth running locally for say 1000 times (ctest can do this automatically) in case you haven't done that to try and catch some more of these types of issues. I'm a bit worried that we might merge this and then every other PR fails in this test :)

Hi @nilsdeppe - there is one failure that I need some help fixing. The unit test helper class for boundary conditions `TestHelpers::evolution::dg::test_boundary_condition_with_python` fails with the message below saying that it does not elicit an error message from c++ for a test case that should:

```
REQUIRE( python_error_message.has_value() == error_msg.has_value() )
with expansion:
  true == false
with messages:
  Seed is: 2537998662 from /__w/spectre/spectre/tests/Unit/Evolution/Systems/
  GeneralizedHarmonic/BoundaryConditions/Test_Bjorhus.cpp:33
  bc_string := "ConstraintPreservingPhysical"
  Test boundary condition after serialization.
  FaceDim := 0
  python_module := "Evolution.Systems.GeneralizedHarmonic.BoundaryConditions.
  Bjorhus"
  face_points := ()
  use_moving_mesh := true
  epsilon := 0.00001
  pretty_type::short_name<BoundaryCorrection>() := "UpwindPenalty"
  python_error_msg_function := "error"
  python_error_message.value_or("") := "Incoming characteristic speeds for
  constraint preserving radiation boundary. Its unclear if proper boundary
  conditionsare imposed in this case."
  error_msg.value_or("") := ""
```

Any ideas would be helpful!

geoffrey4444 commented:
We took a quick look on the call, and we suggest putting in print statements to figure out why c++ and python are doing different things...maybe the zero speed mode is incoming in one but not the other because of a bug?

prayush commented:
@nilsdeppe @geoffrey4444 - this should be good to go now.

nilsdeppe commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/Test_Bjorhus.cpp:
```diff
@@ -0,0 +1,193 @@
+                  "[Unit][Evolution]") {
+  pypp::SetupLocalPythonEnvironment local_python_env{""};
+
+  //   test<1>();
```
why is the 1d case commented out?

prayush replied:
Re-enabled it. There was a rarely triggered [bug](https://github.com/sxs-collaboration/spectre/pull/3172#issuecomment-864992881) that I hadn't caught yet, but I found and fixed it.

markscheel commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/Bjorhus.hpp:
```diff
@@ -0,0 +1,287 @@
+ * written down in Eq. (63) - (65) of \cite Lindblom2005qh . The gauge degrees
+ * of freedom are controlled by imposing a Sommerfeld-type condition (\f$L=0\f$
+ * member of the hierarchy derived in \cite BaylissTurkel) that allow gauge
+ * perturvations to pass through the boundary without strong reflections. These
```
[nitpick] spelling

prayush replied:
fixed

markscheel commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/Bjorhus.hpp:
```diff
@@ -0,0 +1,287 @@
+ * the system that are proportional to the Newman-Penrose curvature spinor
+ * components \f$\Psi_4\f$ and \f$\Psi_0\f$. Here we use Eq. (68) of
+ * \cite Lindblom2005qh to only disallow any incoming waves.
+ *
```
Should it be stated that for char speed precisely equal to zero, that the BC is imposed?

prayush replied:
sure, added a sentence to that effect.

geoffrey4444 commented on src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/Bjorhus.cpp:
```diff
@@ -0,0 +1,469 @@
+    const auto negative_lambda0 =
+        dot_product(normal_covector, *face_mesh_velocity);
+    if (min(-get(negative_lambda0)) < 0) {
+      return {
```
Why doesn't this check the char speed? Why check only negative_lambda0?

prayush replied:
I added this with the intention of catching the case when the outer boundary is being made to move _outward_, and that is why I am only checking the projection of `face_mesh_velocity` onto the face normal. In case the interface is moving outward we would have to extrapolate data inside of the domain to account for the increasing domain volume which is probably a scenario best avoided. 

Char speeds could legitimately be both positive and negative, so why would we want to return an error message for that?

Edit: Rephrased the error message better to indicate this.

nilsdeppe replied:
okay, what about renaming the `negative_lambda0` to `radial_mesh_velocity`? Currently it sounds like it's related to the char speeds, which it sort of is, but that's not why you're computing it.

Can `min(-get(...))` be changed to `max(get(...))`? Would that provide the same check? I find the extra negative with the min makes it a bit more difficult to reason about the check. 

 It might also be a good idea to add the `radial_mesh_velocity` to the error message so users can get some idea of what is going on. What do you think?

Edit: if you're renaming and changing the checks, the `0` -> `0.0` since it's a double

prayush replied:
made all these changes, i.e. renamed `negative_lambda0` to `radial_mesh_velocity`, changed `min` to `max`, updated the error message and switched `0` to `0.`.

prayush replied:
Question for @geoffrey4444 : Since this error was getting triggered for your rotating grid, which should have only tangential motion, should we change the condition here from `max(get(radial_mesh_velocity)) > 0` to `max(get(radial_mesh_velocity)) > 1.e-12` ?

geoffrey4444 replied:
I think that makes sense, and I think that depending on what the BBHs do, the tolerance might need to be adjusted down the road

prayush commented:
@geoffrey4444 @nilsdeppe @markscheel : all fixes are in and all unit tests are passing.

moxcodes reviewed (commented):
fixups look good to me -- I'm happy for you to squash once the other reviewers have a chance to have a look.

nilsdeppe reviewed (commented):
I'm happy for you to squash the fixups

geoffrey4444 commented:
I'm also happy for you to squash

prayush committed:
Update tolerance for radial mesh velocity in Bjorhus class

When we have grids that are tangentially moving at the outer
boundary, e.g. uniformly rotating spheres, the normal mesh
velocity may get a floating-point error level outgoing
velocity. As we do not want this to throw the error that is
supposed to be thrown when the user genuinely uses an outward
moving outer boundary, we update the tolerance for triggering
of that error.

prayush commented:
@sxs-collaboration/spectre-core-devs - this one's probably good to go now.
